scipy/optimize/zeros.py

```diff
@@ -244,6 +264,96 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     return _results_select(full_output, (p, funcalls, itr + 1, _ECONVERR))
 
 
+def _array_newton(func, x0, fprime, args, tol, maxiter, fprime2,
+                  failure_idx_flag=False):
+    """
+    A vectorized version of Newton, Halley, and secant methods for arrays. Do
+    not use this method directly. This method is called from :func:`newton`
+    when ``np.isscalar(x0)`` is true. For docstring, see :func:`newton`.
+    """
+    try:
+        p = np.asarray(x0, dtype=float)
+    except TypeError:  # can't convert complex to float
+        p = np.asarray(x0)
+    failures = np.ones_like(p, dtype=bool)  # at start, nothing converged
+    if fprime is not None:
+        # Newton-Raphson method
+        for iteration in range(maxiter):
+            fder = np.asarray(fprime(p, *args))
+            nz_der = (fder != 0)
+            # stop iterating if all derivatives are zero
+            if not nz_der.any():
```

Comment:
What is the result if the derivative is 0 at the root? E.g. f(x)=x**3-x**2, starting at x0=0 (immediately at the root),  or x0=0.5 (one NR step sends x0 to x1=0.)  It looks like this might be considered a failure?

Reply:
But this is exactly the same code and outcome for zero derivatives as for the scalar case on [lines 196-200](https://github.com/scipy/scipy/blob/14de485c053cc275682bc7de0316fcca72b341f8/scipy/optimize/zeros.py#L196-200)

Reply:
I agree that it's a bug in both.  Are you planning on addressing either as part of this PR?

Reply:
IMO this should be addressed in a separate PR. Are you planning to create a GH issue to track it?

Reply:
@mikofski I understand you eagerness to get this PR in, it's been open for quite a while. It still would be best to address the problem first though; merging something with a known bug and then immediately opening an issue to track the bug is not the best way of going about things from my point of view. Could you try and address this?

Reply:
Hi @rgommers , I created #8904, but now I'm not sure what outcome we expect. Although both the scalar and vectorized versions do raise a `RuntimeWarning` they also return the correct value of zero. So is this a really a bug? Do we _not_ want `newton` to return a warning of an infinite Newton step? I'm not convinced this is a bug, I think it's a feature.

Reply:
I am implementing this change now - pls wait - thx

Reply:
Hi @rgommers and @pvanmulbregt , IMHO #8904 should be handled in a separate PR, and I don't think it should block this PR. Making changes to the what is considered `RuntimeWarning` is IMO a larger issue that should be addressed by more members of the SciPy community. I believe addressing them in this PR may possibly create confusion later on as to why or when the philosophy was changed for when we consider an infinite Newton step a convergence error. I discuss this in #8904 and [I've posted a question on the SciPy mailing list](https://mail.python.org/pipermail/scipy-dev/2018-June/022861.html).